Round one of refactors/tweaks

DirectXMan12 opened the pull request:
This separates code out a bit for readability (IMO), fixes up the importer to use a hybrid source/compiled importer to pull all the code in properly, and switches over to actually detecting if a given argument implements the error interface instead of guessing based on name.

DirectXMan12 committed:
Refactor out importer

This importer should work to import files appropriately.  It attempts to
use the built-in importer (which relies on intermediate artifacts) when
possible, and falls back to source loading when not.

DirectXMan12 committed:
Split fixes out into separate packages

Fixes now live in fixes/, while the fixer lives in fixer/

DirectXMan12 committed:
Actually check if type implements error

Instead of guessing if a type implements error based on the name,
we instead load the error interface up, and use the typechecking
information to confirm.  We could hard-code the error interface,
but this future-proofs us in case of changes, etc.

DirectXMan12 committed:
Add additonal godoc comments

thockin commented on importer/importer.go:
```diff
@@ -0,0 +1,182 @@
+		wg.Add(1)
+		go func(ind int, filepath string) {
+			defer wg.Done()
+			files[ind], errors[ind] = parser.ParseFile(i.fileSet, filepath, nil, 0) // ok to access fset concurrently
```
further down main.go we have logic to re-parse the file to update the AST.  That is happening outside of this framework.  That doesn't seem right?

DirectXMan12 replied:
we're just parsing, not typechecking, so it's fine.  Most of this logic is for import traversal and whatnot in the typechecking phase -- just parsing can be done mostly in a vacuum.

Or are you suggesting to change it for organizational purposes?

thockin commented on fixer/fixer.go:
```diff
@@ -0,0 +1,82 @@
+
+func (f *Fixer) FixPackage(pkg *importer.PackageInfo) error {
+	pkgLog := f.Log.WithValues("package", pkg.BuildInfo.ImportPath)
+	pkgLog.V(1).Info("applying fixes", "file count", len(pkg.Files))
```
Shouldn't logr keys be single tokens by convention?

DirectXMan12 replied:
I've been using space-separated by convention elsewhere.  It doesn't make a whole ton of difference.

thockin commented on fixes/logr.go:
```diff
@@ -190,14 +197,14 @@ func tryPkgExprCall(pkgName string, cursor *astutil.Cursor) bool {
 	// All of these could be embedded in larger expressions.
 	switch selexpr.Sel.Name {
 	case "Info", "Infof", "Infoln", "Warning", "Warningf", "Warningln":
 		fixInfo(selexpr, callexpr)
```
why not a method call here?

thockin commented on main.go:
```diff
@@ -7,69 +7,53 @@
 		fmt.Fprintf(os.Stderr, "\t%s\n", desc)
 	}
-	os.Exit(93)
+	os.Exit(93)  // carefully chosen artisinal exit code
```
artisanal

thockin commented on fixes/logr.go:
```diff
@@ -2,7 +2,7 @@
 // license that can be found in the LICENSE file.
 
-package main
+package fixes
```
I was trying to avoid spending a lot of logic on making this a generic framework for fixes.  I was aiming for "bespoke".  That said, most of this is reasonable cleanup.

If we're having "fixers" I would just as soon have an interface and put each fixer in its own pkg, so some of the method-call conversions are not needed...

DirectXMan12 replied:
that's fine by be -- I mainly just kept this structure while refactoring to avoid too many changes at once.  Alternatively, just getting rid of the "general" fixes pattern is fine.

This is mainly a case of "I saw your TODO about getting rid of 'fixes', but wanted to tackle that in a follow up"

thockin commented on fixes/logr.go:
```diff
@@ -339,16 +390,8 @@ func (f *logrFix) fixError(selexpr *ast.SelectorExpr, callexpr *ast.CallExpr) {
 	for i, arg := range callexpr.Args {
 		t := f.loader.TypeInfo().Types[arg].Type
 		f.log.V(5).Info("arg", "idx", i, "type", t.String())
-		if t.String() == "error" {
```
I think it might still be valuable to check this.  If I have 2 arguments, and one is `error` type and the other implements `error`, I'd prefer to choose the typed one.

DirectXMan12 replied:
Might also be worth checking what golang.org/x/exp/errors/fmt does (the go 2 errors prototype).  I think it uses the last one, but I only glanced at the source code.

DirectXMan12 replied:
(agreed on that point though)

thockin reviewed (commented):
This is basically a rewrite :)  It seems like it shouldn't be this hard -- Go is supposed to great for these types of tools....

DirectXMan12 commented:
>  It seems like it shouldn't be this hard

yeah... convincing Go to load from source is strangely a bit tough.  This loading scenario is what some of the internal tools do (e.g. the Go API checking tool, more-or-less), though.  Others (gotype) skip loading from source and have all sorts of weird behavior because of it.  It would be nice to have a standard form of this.

DirectXMan12 committed:
Switch to tools/packages for loading

tools/packages is module aware, and does a lot of nice stuff for us
(e.g. support for ...).  It's supposed to be the future of
parsing/typechecking/etc libraries in Go.

DirectXMan12 commented:
pushed updates to use the new module-aware `x/tools/go/packages`

thockin commented:
I'm not going to have time in the near future to really comprehend this. :(

Maybe you should just fork it and run with your fork.  I am happy to talk about the theory I had for how to evolve this tool...

DirectXMan12 commented:
that's fine -- just wanted to get the code up for others to take a look at.  Can totally just run with my fork :-)
